pkg/blockbuilder/scheduler/scheduler.go

```diff
@@ -78,24 +86,91 @@ func (s *BlockBuilderScheduler) running(ctx context.Context) error {
 
 func (s *BlockBuilderScheduler) updateSchedule(ctx context.Context) {
 	startTime := time.Now()
-	// Eventually this will also include job computation. But for now, collect partition data.
-	admin := kadm.NewClient(s.kafkaClient)
+	defer func() {
+		s.metrics.updateScheduleDuration.Observe(time.Since(startTime).Seconds())
+	}()
+
+	// TODO: Commit the offsets back to Kafka if dirty.
 
-	startOffsets, err := admin.ListStartOffsets(ctx, s.cfg.Kafka.Topic)
+	lag, err := blockbuilder.GetGroupLag(ctx, s.adminClient, s.cfg.Kafka.Topic, s.cfg.BuilderConsumerGroup, 0)
 	if err != nil {
-		level.Warn(s.logger).Log("msg", "failed to list start offsets", "err", err)
+		level.Warn(s.logger).Log("msg", "failed to get group lag", "err", err)
+		return
+	}
+
+	if ps, ok := lag[s.cfg.Kafka.Topic]; ok {
+		for part, gl := range ps {
+			partStr := fmt.Sprint(part)
+			s.metrics.partitionStartOffset.WithLabelValues(partStr).Set(float64(gl.Start.Offset))
+			s.metrics.partitionEndOffset.WithLabelValues(partStr).Set(float64(gl.End.Offset))
+			s.metrics.partitionCommittedOffset.WithLabelValues(partStr).Set(float64(gl.Commit.At))
+		}
 	}
-	endOffsets, err := admin.ListEndOffsets(ctx, s.cfg.Kafka.Topic)
+
+	oldTime := time.Now().Add(-s.cfg.ConsumeInterval)
+	oldOffsets, err := s.adminClient.ListOffsetsAfterMilli(ctx, oldTime.UnixMilli(), s.cfg.Kafka.Topic)
 	if err != nil {
-		level.Warn(s.logger).Log("msg", "failed to list end offsets", "err", err)
+		level.Warn(s.logger).Log("msg", "failed to obtain old offsets", "err", err)
+		return
 	}
 
-	s.metrics.updateScheduleDuration.Observe(time.Since(startTime).Seconds())
+	// See if the group-committed offset per partition is behind our "old" offsets.
 
-	startOffsets.Each(func(o kadm.ListedOffset) {
-		s.metrics.partitionStartOffset.WithLabelValues(fmt.Sprint(o.Partition)).Set(float64(o.Offset))
-	})
-	endOffsets.Each(func(o kadm.ListedOffset) {
-		s.metrics.partitionEndOffset.WithLabelValues(fmt.Sprint(o.Partition)).Set(float64(o.Offset))
+	oldOffsets.Each(func(o kadm.ListedOffset) {
+		if l, ok := lag.Lookup(o.Topic, o.Partition); ok {
+			if l.Commit.At < o.Offset {
+				level.Info(s.logger).Log("msg", "partition ready", "p", o.Partition)
```

Comment:
From the current block-builder's experience, having a log record, with the partitions, that were skipped due to no lag, was helpful to understand what had happened during an unexpected behaviour.